File: sdk/go/softwarepoe.go
// Package softwarepoe is the Go client SDK for Software POE
// (post-occupancy evaluation for software architecture).
// It mirrors schemas/event.schema.json version 0.1.0.
package softwarepoe

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

const (
	SchemaVersion = "0.1.0"
	Project       = "software-poe"
)

// Metrics is the registered metric ID set.
var Metrics = map[string]bool{
	"intent_fulfillment_ratio": true,
	"slo_gap":                  true,
	"cost_estimate_error":      true,
	"unused_capability_ratio":  true,
	"incident_delta":           true,
}

type Event struct {
	TenantID      string         `json:"tenant_id"`
	Project       string         `json:"project"`
	Source        string         `json:"source"`
	ObservedAt    string         `json:"observed_at"`
	Metric        string         `json:"metric"`
	Value         float64        `json:"value"`
	Unit          string         `json:"unit,omitempty"`
	EvidenceRef   string         `json:"evidence_ref,omitempty"`
	SchemaVersion string         `json:"schema_version"`
	Attributes    map[string]any `json:"attributes,omitempty"`
}

// Validate checks the event against the 0.1.0 contract. Zero-valued
// Project/SchemaVersion fields are filled with the supported constants.
func (e *Event) Validate() []string {
	if e.Project == "" {
		e.Project = Project
	}
	if e.SchemaVersion == "" {
		e.SchemaVersion = SchemaVersion
	}
	var errs []string
	if e.TenantID == "" {
		errs = append(errs, "tenant_id is required")
	}
	if e.Project != Project {
		errs = append(errs, "project must be '"+Project+"'")
	}
	if e.Source == "" {
		errs = append(errs, "source is required")
	}
	if _, err := time.Parse(time.RFC3339, e.ObservedAt); err != nil {
		errs = append(errs, "observed_at must be an RFC3339 date-time")
	}
	if !Metrics[e.Metric] {
		errs = append(errs, "metric must be a registered metric ID")
	}
	if e.SchemaVersion != SchemaVersion {
		errs = append(errs, "schema_version must be '"+SchemaVersion+"'")
	}
	for k, v := range e.Attributes {
		switch v.(type) {
		case string, float64, int, bool, nil:
		default:
			errs = append(errs, "attributes."+k+" must be a scalar")
		}
	}
	return errs
}

type Assessment struct {
	AssessmentID        string   `json:"assessment_id"`
	TenantID            string   `json:"tenant_id"`
	Status              string   `json:"status"`
	Summary             string   `json:"summary"`
	Uncertainty         float64  `json:"uncertainty"`
	GapDetected         bool     `json:"gap_detected"`
	MissingData         bool     `json:"missing_data"`
	StaleEvidence       bool     `json:"stale_evidence"`
	RuleVersion         *string  `json:"rule_version"`
	EvidenceRefs        []string `json:"evidence_refs"`
	HumanReviewRequired bool     `json:"human_review_required"`
	CreatedAt           string   `json:"created_at"`
}

type Client struct {
	baseURL string
	apiKey  string
	http    *http.Client
}

func NewClient(baseURL, apiKey string) *Client {
	return &Client{
		baseURL: baseURL,
		apiKey:  apiKey,
		http:    &http.Client{Timeout: 10 * time.Second},
	}
}

func (c *Client) do(method, path string, body any, out any) error {
	var reader io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(buf)
	}
	req, err := http.NewRequest(method, c.baseURL+path, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("X-API-Key", c.apiKey)
	}
	resp, err := c.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		msg, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("http %d: %s", resp.StatusCode, msg)
	}
	if out == nil {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

type IngestResponse struct {
	EventID     string `json:"event_id"`
	EvidenceRef string `json:"evidence_ref"`
	Status      string `json:"status"`
}

func (c *Client) Ingest(event *Event) (*IngestResponse, error) {
	if errs := event.Validate(); len(errs) > 0 {
		return nil, fmt.Errorf("invalid event: %v", errs)
	}
	var out IngestResponse
	if err := c.do(http.MethodPost, "/v1/software-poe/events", event, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *Client) Assessments(tenantID, status string) ([]Assessment, error) {
	q := url.Values{"tenant_id": {tenantID}}
	if status != "" {
		q.Set("status", status)
	}
	var out []Assessment
	if err := c.do(http.MethodGet, "/v1/software-poe/assessments?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

type ReviewRequest struct {
	Reviewer   string `json:"reviewer"`
	Decision   string `json:"decision"`
	Annotation string `json:"annotation,omitempty"`
}

func (c *Client) Review(assessmentID, tenantID string, review ReviewRequest) error {
	q := url.Values{"tenant_id": {tenantID}}
	path := "/v1/software-poe/assessments/" + assessmentID + "/review?" + q.Encode()
	return c.do(http.MethodPost, path, review, nil)
}

File: sdk/julia/src/SoftwarePOE.jl
"""
    SoftwarePOE

Client SDK for Software POE (post-occupancy evaluation for software
architecture). Mirrors `schemas/event.schema.json` version 0.1.0.
"""
module SoftwarePOE

using Dates, HTTP, JSON3, StructTypes

export Client, Event, validate_event, ingest, assessments, review

const SCHEMA_VERSION = "0.1.0"
const PROJECT = "software-poe"
const METRICS = Set([
    "intent_fulfillment_ratio",
    "slo_gap",
    "cost_estimate_error",
    "unused_capability_ratio",
    "incident_delta",
])

Base.@kwdef struct Event
    tenant_id::String
    source::String
    observed_at::String
    metric::String
    value::Float64
    unit::Union{String,Nothing} = nothing
    evidence_ref::Union{String,Nothing} = nothing
    attributes::Union{Dict{String,Any},Nothing} = nothing
    project::String = PROJECT
    schema_version::String = SCHEMA_VERSION
end

StructTypes.StructType(::Type{Event}) = StructTypes.Struct()
StructTypes.omitempties(::Type{Event}) = (:unit, :evidence_ref, :attributes)

"""Return a vector of validation error strings (empty when valid)."""
function validate_event(e::Event)
    errors = String[]
    isempty(e.tenant_id) && push!(errors, "tenant_id is required")
    e.project == PROJECT || push!(errors, "project must be '$(PROJECT)'")
    isempty(e.source) && push!(errors, "source is required")
    try
        DateTime(replace(e.observed_at, r"Z$" => ""), dateformat"yyyy-mm-dd\THH:MM:SS")
    catch
        push!(errors, "observed_at must be an RFC3339 date-time")
    end
    e.metric in METRICS || push!(errors, "metric must be a registered metric ID")
    e.schema_version == SCHEMA_VERSION ||
        push!(errors, "schema_version must be '$(SCHEMA_VERSION)'")
    if e.attributes !== nothing
        for (k, v) in e.attributes
            v isa Union{AbstractString,Real,Bool,Nothing} ||
                push!(errors, "attributes.$k must be a scalar")
        end
    end
    return errors
end

struct Client
    base_url::String
    headers::Vector{Pair{String,String}}
end

function Client(base_url::AbstractString; api_key::Union{AbstractString,Nothing}=nothing)
    headers = ["Content-Type" => "application/json"]
    api_key !== nothing && push!(headers, "X-API-Key" => String(api_key))
    Client(rstrip(String(base_url), '/'), headers)
end

function ingest(c::Client, e::Event)
    errors = validate_event(e)
    isempty(errors) || throw(ArgumentError("invalid event: " * join(errors, "; ")))
    resp = HTTP.post(c.base_url * "/v1/software-poe/events", c.headers, JSON3.write(e))
    return JSON3.read(resp.body)
end

function assessments(c::Client, tenant_id::AbstractString; status=nothing)
    query = Dict("tenant_id" => tenant_id)
    status !== nothing && (query["status"] = status)
    resp = HTTP.get(c.base_url * "/v1/software-poe/assessments", c.headers; query=query)
    return JSON3.read(resp.body)
end

function review(c::Client, assessment_id::AbstractString, tenant_id::AbstractString;
                reviewer::AbstractString, decision::AbstractString, annotation=nothing)
    body = JSON3.write(Dict(
        "reviewer" => reviewer, "decision" => decision, "annotation" => annotation))
    url = c.base_url * "/v1/software-poe/assessments/$(assessment_id)/review"
    resp = HTTP.post(url, c.headers, body; query=Dict("tenant_id" => tenant_id))
    return JSON3.read(resp.body)
end

end # module
